core/primitives/src/reed_solomon.rs

```diff
@@ -0,0 +1,85 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use itertools::Itertools;
+use reed_solomon_erasure::galois_8::{Field, ReedSolomon};
+use reed_solomon_erasure::ReconstructShard;
+use std::io::Error;
+
+/// The ttl for a reed solomon instance to control memory usage. This number below corresponds to
+/// roughly 60MB of memory usage.
+const RS_TTL: u64 = 2 * 1024;
+
+/// Wrapper around reed solomon which occasionally resets the underlying
+/// reed solomon instead to work around the memory leak in reed solomon
+/// implementation <https://github.com/darrenldl/reed-solomon-erasure/issues/74>
+pub struct ReedSolomonWrapper {
+    rs: ReedSolomon,
+    ttl: u64,
+}
+
+impl ReedSolomonWrapper {
+    pub fn new(data_shards: usize, parity_shards: usize) -> Self {
+        ReedSolomonWrapper {
+            rs: ReedSolomon::new(data_shards, parity_shards).unwrap(),
+            ttl: RS_TTL,
+        }
+    }
+
+    // Encode function takes a serializable object and returns a tuple of parts and length of encoded data
+    pub fn encode<T: BorshSerialize>(&mut self, data: T) -> (Vec<Option<Box<[u8]>>>, usize) {
+        let mut bytes = borsh::to_vec(&data).unwrap();
+        let encoded_length = bytes.len();
+
+        let data_parts = self.rs.data_shard_count();
+        let part_length = (encoded_length + data_parts - 1) / data_parts;
+
+        // Pad the bytes to be a multiple of `part_length`
+        // Convert encoded data into `data_shard_count` number of parts and pad with `parity_shard_count` None values
+        // with 4 data_parts and 2 parity_parts
+        // b'aaabbbcccd' -> [Some(b'aaa'), Some(b'bbb'), Some(b'ccc'), Some(b'd00'), None, None]
+        bytes.resize(data_parts * part_length, 0);
+        let mut parts = bytes
+            .chunks_exact(part_length)
+            .map(|chunk| Some(chunk.to_vec().into_boxed_slice()))
+            .chain(itertools::repeat_n(None, self.rs.parity_shard_count()))
+            .collect_vec();
+
+        // Fine to unwrap here as we just constructed the parts
+        self.reconstruct(&mut parts).unwrap();
```

Comment:
Previously if reconstruction failed we were printing a warn and returning. If we abort instead will that be handled appropriately up the call chain? Or, did you realize that this always succeeds and the warn on err is unnecessary?

Reply:
So the unwrap here is totally fine as we just constructed the parts above and we basically guarantee they are properly constructed and in place. The reconstruct here is basically just to populate the parity shards (all the `None`s)

Reply:
The same reconstruct function is called in the decode function where we do need to handle and expose the errors.

Reply:
I see, so encode followed by decode always succeeds?

Reply:
encode does the following: take the whole object, serialize it, split it into parts, populate the parity parts (here we call self.reconstruct) and then return the parts.

decode meanwhile does the following: take a subset of parts, try to reconstruct the data (this also calls self.reconstruct but can fail) and then deserialize and return object (deserialization can also fail)

Reply:
But yes, encode followed by decode always succeeds as long as we have "num_data_parts" number of parts.